### PR TITLE
Fix Tox configuration to use Poetry hermetically

### PR DESCRIPTION
**Pull request by JoshuaLeivers**
With the existing setup, Tox could sometimes end up using the Poetry virtualenv for all test runs. This has now been fixed.

While fixing this, another problem was found, in that the current usage of Pydantic is not compatible with Pydantic v2 and will raise errors in some situations, specifically including test_oneof.py. Pydantic has now been limited to v1 as a result.

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+envlist = py37, py38, py310
+isolated_build = true
+
+[testenv]
+require_poetry = true
+require_locked_deps = true
+poetry_dep_groups =
+    dev
+commands =
+    poe generate
```
Comment (a-khabarov) [inline tox.ini]: I don't think we should be making this part of the `tox` commands - `poe generate` can be quite slow. Also, is there a reason why this is now a separate file? Probably best not to make it a separate file if we can.

File: pyproject.toml [hunk 2]
```diff
@@ -37,7 +37,7 @@ sphinx-rtd-theme = "0.5.0"
 tomlkit = "^0.7.0"
 tox = "^3.15.1"
 pre-commit = "^2.17.0"
-pydantic = ">=1.8.0"
+pydantic = ">=1.8.0,<2"
```
Comment (a-khabarov) [inline pyproject.toml]: This should probably be a separate commit?

File: pyproject.toml [hunk 3]
```diff
@@ -57,6 +57,10 @@ help = "Generate test cases (do this once before running test)"
 cmd = "pytest"
 help = "Run tests"
 
+[tool.poe.tasks.test-coverage]
```
Comment (a-khabarov) [inline pyproject.toml]: I don't think we need this, we can just have `pytest --cov betterproto` in the `tox` config.

File: tox.ini (hunk 1 above)
Comment (JoshuaLeivers) [reply tox.ini]: I have now removed the `generate` call. You are right that it is slow and adds significant time to the runs, though I am somewhat concerned that this means that some Python-version-specific errors in compilation may not be caught. I did see behaviour that would suggest this might actually be the case when it came to the Pydantic v2 incompatibility.

File: pyproject.toml (hunk 2 above)
Comment (JoshuaLeivers) [reply pyproject.toml]: This has now been moved to a separate commit

File: pyproject.toml (hunk 3 above)
Comment (JoshuaLeivers) [reply pyproject.toml]: This has now been removed and Tox no longer uses Poe, as it seems to use the Poetry Virtualenv no matter what, which defeats the point of this fix.

File: tox.ini [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    py310
+    py38
+
+[testenv]
+commands =
+    pytest --cov betterproto
+poetry_dep_groups =
+    dev
+require_locked_deps = true
+require_poetry = true
```
Comment (mattyclarkson) [inline tox.ini]: I would suggest the following.

- Keep `isolated_build`
- Install the tox plugin as part of `tox`, then `poetry` is resolved hermetically

```suggestion
[tox]
requires =
    tox-poetry-installer[poetry]==1.0.0b1
    tox>=4.2
envlist =
    py3
env_list =
    py311
    py310
    py39
    py38
    py37
isolated_build = true
[testenv]
require_poetry = true
require_locked_deps = true
poetry_dep_groups =
    dev
commands =
    pytest --cov {posargs} betterproto
```

File: pyproject.toml [hunk 5]
```diff
@@ -34,10 +34,11 @@ pytest-cov = "^2.9.0"
...
 sphinx = "3.1.2"
 sphinx-rtd-theme = "0.5.0"
-tomlkit = "^0.7.0"
-tox = "^3.15.1"
+tomlkit = ">=0.7.0"
+tox = "^4.0.0"
 pre-commit = "^2.17.0"
-pydantic = ">=1.8.0"
+pydantic = "^1.8.0"
+tox-poetry-installer = {version = "1.0.0b1", extras = ["poetry"]}
```
Comment (mattyclarkson) [inline pyproject.toml]: I would recommend installing this through `tox` because otherwise you need `poetry` to install the plugin which then uses poetry. Weird.

File: pyproject.toml [hunk 6]
```diff
@@ -12,15 +12,15 @@ packages = [
...
+python = "^3.8"
 black = { version = ">=23.1.0", optional = true }
 grpclib = "^0.4.1"
 importlib-metadata = { version = ">=1.6.0", python = "<3.8" }
 jinja2 = { version = ">=3.0.3", optional = true }
 python-dateutil = "^2.8"
 isort = {version = "^5.11.5", optional = true}
 
-[tool.poetry.dev-dependencies]
+[tool.poetry.group.dev.dependencies]
```
Comment (mattyclarkson) [inline pyproject.toml:23]: I would usually recommend splitting between dev dependencies and test dependencies:

- `dev` tools for woring with the project
- `test` packages that should be installed in the test (`tox`) environment

This can be important because when testing the project `tox` will install all the dependencies when they aren't all needed. Also, `tox` can self-provision itself so if someone has `tox` installed and we can resolve `poetry` hermetically they can test the project without ever having `poetry` installed. They just run `tox`.

File: pyproject.toml [hunk 7]
```diff
@@ -12,15 +12,15 @@ packages = [
 ]
 
 [tool.poetry.dependencies]
-python = "^3.7"
+python = "^3.8"
```
Comment (mattyclarkson) [inline pyproject.toml]: This is quite an important change as it changes the Python version the project supports. I _assume_ one of the dependency bumps requires this?

**Review by mattyclarkson — CHANGES_REQUESTED** [review]

File: tox.ini (hunk 4 above)
Comment (JoshuaLeivers) [reply tox.ini]: I don't have a problem with keeping `isolated_build`, but I'm not sure what the benefit is of specifying the requirement inside `tox.ini` rather than `pyproject.toml`. The `tox-poetry-installer` docs recommend against it, but I'm not sure what the behaviour difference would be. Having it in `pyproject.toml` has caught the fact that the project requires `python>=3.7` but `tox-poetry-installer` doesn't support 3.7, which having it in `tox.ini` seemed to not flag.

File: pyproject.toml (hunk 6 above)
Comment (JoshuaLeivers) [reply pyproject.toml:23]: I'll have a look at splitting them up, good point

File: pyproject.toml (hunk 7 above)
Comment (JoshuaLeivers) [reply pyproject.toml]: Yes, `tox-poetry-installer` does not support 3.7, it seems. I've mentioned it in the upstream issue, and will ask about it if we get an upstream PR and that hasn't had a response. I saw mention on Slack of them potentially dropping 3.7 at some point, so the maintainers don't seem _completely_ adverse to the idea, will just have to get their input.

File: pyproject.toml (hunk 6 above)
Comment (JoshuaLeivers) [reply pyproject.toml:23]: This has now been done

File: tox.ini (hunk 4 above)
Comment (JoshuaLeivers) [reply tox.ini]: It appears that `isolated_build` is no longer an option, as the behaviour is now the default. https://tox.wiki/en/latest/upgrading.html#isolated-environment-on-by-default

File: pyproject.toml [hunk 8]
```diff
@@ -12,33 +12,35 @@ packages = [
...
+sphinx = "3.1.2"
+sphinx-rtd-theme = "0.5.0"
+tomlkit = ">=0.7.0"
+pre-commit = "^2.17.0"
+grpcio-tools = "^1.54.2"
+
+[tool.poetry.group.test.dependencies]
 poethepoet = ">=0.9.0"
-protobuf = "^4.21.6"
+tox = "^4.0.0"
```
Comment (mattyclarkson) [inline pyproject.toml]: This is a `dev` dependency.

File: tox.ini [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+requires =
+    tox>=4.2
+env_list =
+    py311
+    py310
+    py38
+
+[testenv]
+commands =
+    pytest --cov betterproto
```
Comment (mattyclarkson) [inline tox.ini]:
```suggestion
    pytest {posargs:--cov betterproto}
```

This is _really_ useful to override the arguments passed to `pytest`. It means you can run just one test with:

```
tox -e py -- betterproto/test_abc.py
```

**Review by mattyclarkson — CHANGES_REQUESTED** [review]
